src/coreclr/jit/optcse.cpp

```diff
@@ -4963,325 +4832,308 @@ void CSE_HeuristicCommon::PerformCSE(CSE_Candidate* successfulCandidate)
-        var_types expTyp = genActualType(exp->TypeGet());
-
-        // The cseLclVarType must be a compatible with expTyp
-        //
-        ValueNumStore* vnStore = m_pCompiler->vnStore;
-        noway_assert(IsCompatibleType(cseLclVarTyp, expTyp) || (dsc->csdConstDefVN != vnStore->VNForNull()));
-
-        // This will contain the replacement tree for exp
-        // It will either be the CSE def or CSE ref
-        //
-        GenTree* cse = nullptr;
-        bool     isDef;
+#ifdef DEBUG
+        if (m_pCompiler->verbose)
+        {
+            printf("\n" FMT_CSE " def at ", GET_CSE_INDEX(exp->gtCSEnum));
+            Compiler::printTreeID(exp);
+            printf(" replaced in " FMT_BB " with def of V%02u\n", blk->bbNum, cseLclVarNum);
+        }
+#endif // DEBUG
 
-        if (IS_CSE_USE(exp->gtCSEnum))
+        GenTree* val = exp;
+        if (isSharedConst)
         {
-            /* This is a use of the CSE */
-            isDef = false;
-#ifdef DEBUG
-            if (m_pCompiler->verbose)
+            ValueNum currVN   = m_pCompiler->vnStore->VNLiberalNormalValue(exp->gtVNPair);
+            ssize_t  curValue = m_pCompiler->vnStore->CoercedConstantValue<ssize_t>(currVN);
+            ssize_t  delta    = curValue - dsc->csdConstDefValue;
+            if (delta != 0)
             {
-                printf("\nWorking on the replacement of the " FMT_CSE " use at ", exp->gtCSEnum);
-                Compiler::printTreeID(exp);
-                printf(" in " FMT_BB "\n", blk->bbNum);
+                val = m_pCompiler->gtNewIconNode(dsc->csdConstDefValue, cseLclVarTyp);
+                val->gtVNPair.SetBoth(dsc->csdConstDefVN);
             }
-#endif // DEBUG
+        }
 
-            // We will replace the CSE ref with a new tree
-            // this is typically just a simple use of the new CSE LclVar
-            //
+        // Create a store of the value to the temp
+        GenTree* store     = m_pCompiler->gtNewTempStore(cseLclVarNum, val);
+        GenTree* origStore = store;
 
-            // Create a reference to the CSE temp
-            GenTreeLclVar* cseLclVar = m_pCompiler->gtNewLclvNode(cseLclVarNum, cseLclVarTyp);
-            cseLclVar->gtVNPair.SetBoth(dsc->csdConstDefVN);
+        if (!store->OperIs(GT_STORE_LCL_VAR))
+        {
+            // This can only be the case for a struct in which the 'val' was a COMMA, so
+            // the store is sunk below it.
+            store = store->gtEffectiveVal();
+            noway_assert(origStore->OperIs(GT_COMMA) && (origStore == val));
+        }
+        else
+        {
+            noway_assert(store->Data() == val);
+        }
 
-            cse = cseLclVar;
-            if (isSharedConst)
+        // Assign the proper Value Numbers.
+        ValueNumPair valExc = m_pCompiler->vnStore->VNPExceptionSet(val->gtVNPair);
+        store->gtVNPair     = m_pCompiler->vnStore->VNPWithExc(ValueNumStore::VNPForVoid(), valExc);
```

Comment:
This was assigning `ValueNumStore::VNPForVoid` before, without any exceptions. That seems wrong.